Move Environment and friends to corefx

stephentoub opened the pull request:
Currently System.Runtime.Extensions type forwards to the Environment in the runtime, which means we need to implement functionality in both coreclr and corert.  It's also missing functionality, which means to add that functionality back we'd need to do so in both runtimes, and we'd miss out on sharing code with elsewhere in corefx.

This commit:
1) Adds Environment, EnvironmentVariableTarget, OperatingSystem, and PlatformID to corefx so that System.Runtime.Extensions contains the type rather than forwarding to it.
2) Adds implementations of all of the missing members: CommandLine, CurrentDirectory, ExitCode, Is64BitProcess, Is64BitOperatingSystem, OSVersion, SystemPageSize, UserInteractive, UserName, UserDomainName, Version, WorkingSet, GetLogicalDrives, GetFolderPath, and the overloads of Get/SetEnvironmentVariable(s) that take an EnvironmentVariableTarget.
3) Adds the additional members to the contract.
4) Adds some basic tests to ensure basic functionality works.

Work that still needs to be done after this:
1) Add an EnvironmentAugments type to System.Private.Corelib in both corefx and corert, use that to replace the reflection-based hack in this commit, then (potentially) remove the current Environment from Corelib's model.xml so that the rewriter will make it internal.
2) Add more support to the Unix implementation of GetFolderPath for more special folders
3) Implement more Windows functionality if/when missing functions become available, e.g. SHGetFolderPath.
I will open issues for each of these prior to merging this commit.

cc: @jkotas, @weshaggard, @danmosemsft, @KrzysztofCwalina, @ianhays 

Fixes #9826 
Fixes #5407
Fixes #4741
Fixes #3568

ianhays commented:
Yay for OSVersion! resolves https://github.com/dotnet/corefx/issues/4741

jkotas commented on src/Common/src/Interop/Windows/mincore/Interop.GetVersionExW.cs:
```diff
@@ -0,0 +1,31 @@
+            public int dwMinorVersion;
+            public int dwBuildNumber;
+            public int dwPlatformId;
+            [MarshalAs(UnmanagedType.ByValTStr, SizeConst = 128)]
```
`public fixed char szCSDVersion[128];` would be better - faster, and you can also use `sizeof(OSVERSIONINFOEX)` instead of the slow `Marshal.SizeOf` to initialize the size.

stephentoub replied:
Ok, will change.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.Unix.cs:
```diff
@@ -0,0 +1,205 @@
+            get
+            {
+                // TODO: While this is functional, it would be better performing to put an implementation into System.Native
+                // similar to what's used in the libcoreclr's GetTickCount() implementation.
```
`Environment.TickCount` tends to be used on performance sensitive paths because of it has always been the lowest overhead way to get a time stamp. We should keep an eye on perf with this rewrite - on both Windows and Unix.

stephentoub replied:
> We should keep an eye on perf with this rewrite - on both Windows and Unix.

The CoreCLR implementation is an internal call into the runtime which then turns around just calls ::GetTickCount (then on Unix that's implemented in the libcoreclr PAL).  Is a P/Invoke to the Win32 GetTickCount() measurably slower than an internal call into the runtime that then calls the Win32 function (that's how I interpret your concern)?  I do expect this implementation will be slower on Unix, hence the TODO to essentially port the libcoreclr GetTickCount to System.Native.

jkotas replied:
Yes, I think that the PInvoke overhead will be measurable for GetTickCount. Also, GetTickCount64 may be slower than GetTickCount on 32-bit platforms.

The runtime itself needs implementation of GetTickCount anyway, so we can just use it from there.

There is still overhead of the call forwarding, but it is likely going to be quite a bit less than PInvoke overhead.

stephentoub replied:
Ok, I'll go back to call forwarding it.  I was trying to move out everything that wasn't strictly necessary, but we can always change this in the future if desired.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.Windows.cs:
```diff
@@ -0,0 +1,290 @@
+    {
+        private static readonly Lazy<bool> s_isAppX = new Lazy<bool>(() =>
+        {
+            // TODO: Determine the right way to get at this information from outside of System.Private.Corelib
```
A better alternative may be to do different builds of `System.Runtime.Extensions.dll` for appx vs. console. It is what we are doing for all other custom appx-specific implementations in corefx today.

jkotas replied:
We should be able to share the appx-specific build between .NET Native for UWP and CoreCLR.

stephentoub replied:
Yeah, I went back and forth on that.  I actually had the split implementations initially.  I can go back to it.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.cs:
```diff
@@ -0,0 +1,217 @@
+
+        public static void SetEnvironmentVariable(string variable, string value) => SetEnvironmentVariable(variable, value, EnvironmentVariableTarget.Process);
+
+        public static void SetEnvironmentVariable(string variable, string value, EnvironmentVariableTarget target) => EnvironmentAugments.SetEnvironmentVariable(variable, value, target);
```
I do not think environment should go to runtime. It should be implemented in corefx.

stephentoub replied:
You mean environment variables?  I did this because of the coreclr's cache on Unix for reading all of the environment variables and then ensuring synchronized access to that cache.  We can do similar caching in corefx, but then how would we synchronize access between coreclr's use of environment variables and corefx's? And how would we ensure that any environment variables set by coreclr are then visible to corefx? Or maybe coreclr doesn't do any such setting, and it doesn't need to observe any changes made by corefx, in which case none of this would matter and would could move it without problems?

jkotas replied:
Yes, I meant environment variables.

CoreCLR does not need to observe any changes made by corefx. CoreCLR initialization from environment is basically done during startup, any changes in environment after startup are ignored, and the runtime itself does not have a need to change environment variables.

The cache that avoids races and leaks on Unix should move without problems. It exists just to support BCL.

stephentoub replied:
Ok, good.  I'll work on moving the env var stuff to corefx then. Thanks.

danmosemsft commented on src/System.Runtime.Extensions/src/System/Environment.Unix.cs:
```diff
@@ -0,0 +1,205 @@
+            {
+                // First try with a buffer that should suffice for 99% of cases.
+                string username;
+                const int BufLen = 1024;
```
How do you guys test the fallback path for these things? In the past, I've used #if DEBUG to set the initial buffer to zero in debug only.

stephentoub replied:
Done

stephentoub replied:
This was a good idea, and actually helped find a bug in our native implementation (which I fixed), but I'm going to undo it as it's bumping up against what appears to be a known issue on CentOS:
https://bugs.centos.org/view.php?id=7324
and the associated test is failing.

stephentoub commented:
Thanks for the review. I pushed another commit that addresses the feedback and fixes up additional things.  Main changes:
- Split the Windows implementation across builds for CoreCLR and .NET Native
- Moved all of the environment variables functionality to corefx
- Changed TickCount to be forwarded to the runtime
- Used GetLogicalProcessorInformationEx to determine the processor count

plus a smattering of others.

jkotas commented on src/System.Runtime.Extensions/src/System.Runtime.Extensions.csproj:
```diff
@@ -106,16 +143,42 @@
   </ItemGroup>
   <!-- WINDOWS: CoreCLR -->
   <ItemGroup Condition=" '$(TargetsWindows)' == 'true' And '$(CoreClrOrCorRt)' == 'true'">
+    <Compile Include="System\Environment.CoreCLR.cs" />
```
This is not CoreCLR specific. The conventions used elsewhere are:
- .Windows.cs: Windows-specific implementation, usable for both classic and UWP apps
- .Win32.cs: Implementation specific for classic apps
- .WinRT.cs: Implementation specific for UWP apps

jkotas replied:
In other words, we want this implementation to be usable for CoreRT console apps - replacing what's there today (https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Environment.Win32.cs).

stephentoub replied:
I guess I'm confused when we're choosing to use .NETNative.cs/.CoreCLR.cs vs .Win32.cs/.WinRT.cs (we have a bunch of files in the corefx repo that use the former convention, and I was just following that).  But I'll switch them.

jkotas replied:
There are several different axes:

1. Classic win32 vs. UWP apps (vs. non-Windows OSes). The two Windows appmodels are very much like different OSes because of there are different APIs available and things have to be generally done differently in number of places. I believe that the most common pattern has been to use `.Win32.cs` vs. `.WinRT.cs` to provide different implementations. `System.IO.FileSystem` or  `System.Console` are good examples.

2. JIT vs. full AOT. I do not actually see any good examples. `System.Linq.Expressions` or data contract serialization where this difference is most prelevant are configured via ifdefs, not via alternative file includes.

3. Different surface of System.Private.CoreLib between CoreCLR and CoreRT/.NET Native. The common pattern for this is `.CoreCLR.cs` vs. `.netcore50.cs` or `.netcore50aot.cs`. Hopefully, most of these will go away over time as we work on filling in the missing APIs and converging the implementations.

I know there have been a ton of confusion and inconsistencies about this. We have been getting by so far because of we support just a small subset of combinations (CoreCLR/Classic, CoreCLR/UWP, CoreRT/UWP). We run into it during OSS CoreRT project bring up because it needs a new combination (CoreRT/Classic). We ended up hacking around it because of getting this cleaned up (especially wrt packaging) seemed to be a lot of work.

I am going to open an issue on fixing consistency of the .cs file names at least to make some incremental progress on this.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.Unix.cs:
```diff
@@ -0,0 +1,295 @@
+            byte** environ = Interop.Sys.GetEnviron();
+            if (environ != null)
+            {
+                for (byte** ptr = Interop.Sys.GetEnviron(); *ptr != null; ptr++)
```
No need to call `Interop.Sys.GetEnviron` second time.

stephentoub replied:
Oops, yes, will fix.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.CoreCLR.cs:
```diff
@@ -0,0 +1,406 @@
+            {
+                Debug.Assert(target == EnvironmentVariableTarget.User);
+
+                // TODO #8533: Uncomment/fix when registry APIs available
```
I guess we won't be able to just use the registry contract here to avoid bad cycles?

stephentoub replied:
I'm not sure.  If we pull both this and the registry APIs into System.Runtime, then it'd be available, but otherwise it'd be a bad cycle.  We'll probably just need to P/Invoke directly; potentially we could do some small refactoring on Microsoft.Win32.Registry to make the relevant parts easily sharable.

jkotas commented on src/System.Runtime.Extensions/src/System/Environment.CoreCLR.cs:
```diff
@@ -0,0 +1,406 @@
+
+                // TODO #8533: Uncomment/fix when registry APIs available
+                //using (RegistryKey environmentKey = Registry.CurrentUser.OpenSubKey("Environment", false))
+                //{
```
This implementation is not going to be tree-shaking friendly. If I am just using process environment (the common case), it is going to pull in the bag of rare registry stuff. When we see things like this, we may want to factor them to be tree-shaking friendly. In this case, it would mean avoiding the registry in the transitive closure of the process APIs (it would require a few more methods in the PAL).

Mono BCL tends to have tweaks like this to make their BCL tree-shake well.

stephentoub replied:
Ok, I'll refactor it slightly to keep the Process case in its own method.

jkotas commented:
LGTM so far, modulo comments.

stephentoub commented:
Thanks for all the good feedback, @jkotas. Addressed.

@ianhays, FYI, this also fixes https://github.com/dotnet/corefx/issues/3568.

justinvp commented on src/Common/src/System/Collections/Generic/LowLevelDictionary.IDictionary.cs:
```diff
@@ -0,0 +1,149 @@
+
+        object IDictionary.this[object key]
+        {
+            get { return this[(TKey)key]; }
```
Instead of throwing `KeyNotFoundException`, this should return null if the key isn't found.

stephentoub replied:
Fixed

justinvp commented on src/Common/src/System/Collections/Generic/LowLevelDictionary.IDictionary.cs:
```diff
@@ -0,0 +1,149 @@
+
+        void ICollection.CopyTo(Array array, int index)
+        {
+            int dst = 0;
```
Shouldn't `dst` be initialized to `index`? Also, what about parameter validation?

stephentoub replied:
Fixed the initialization.  I'm not bothering with parameter validation as this code should be temporary.

justinvp commented on src/Common/src/System/Collections/Generic/LowLevelDictionary.IDictionary.cs:
```diff
@@ -0,0 +1,149 @@
+                return _pos < _entries.Length;
+            }
+
+            public void Reset() { throw new NotSupportedException(); }
```
`Hashtable`'s enumerator  supports `Reset` (as does `Dictionary<TKey, TValue>`'s).

danmosemsft replied:
Would it be easier to just switch to regular Dictionary at this point? I don't think there's a layering issue.

stephentoub replied:
> Would it be easier to just switch to regular Dictionary at this point? I don't think there's a layering issue.

I'd be happy to and would like to, but I can't until @KrzysztofCwalina's work lands to combine all of the relevant assemblies.  As it currently stands, System.Runtime.Extensions doesn't have access to it.

But that's also why I haven't done a lot of work on the implementation here around argument validation and the like, because the whole LowLevelDictionary thing should completely go away.

justinvp commented on src/Common/src/System/Collections/Generic/LowLevelDictionary.IDictionary.cs:
```diff
@@ -0,0 +1,149 @@
+
+        bool ICollection.IsSynchronized => false;
+
+        object ICollection.SyncRoot => this;
```
Note: `Hashtable.SyncRoot` returns a lazily initialized (via `Interlocked.CompareExchange`) private object.

justinvp commented on src/Common/src/System/Collections/Generic/LowLevelDictionary.IDictionary.cs:
```diff
@@ -0,0 +1,149 @@
+            set { this[(TKey)key] = (TValue)value; }
+        }
+
+        ICollection IDictionary.Keys
```
`Hashtable.Keys` and `.Values` return read-only collections. Do we care here?

stephentoub replied:
I've not invested a lot of time in this implementation; it's just enough to satisfy the basics of the API.  It should be replaced by Dictionary or Hashtable once those are accessible to this code.

justinvp replied:
FYI, if you do end up switching to Dictionary, you may want to take a look at what I did for `SqlConnection.RetreiveStatistics`, which previously returned Hashtable but now returns Dictionary (to break the dependency on the non-generic collections). I had to subclass Dictionary to make it fully compatible with Hashtable.

https://github.com/dotnet/corefx/blob/a59bf6473733827293030457c8a8b147f1e573af/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlStatistics.cs#L225-L333

stephentoub commented:
Test Innerloop CentOS7.1 Release Build and Test please

stephentoub commented:
Test Innerloop CentOS7.1 Release Build and Test please ("failed to mkdirs")

niemyjski commented on src/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs:
```diff
@@ -534,6 +614,27 @@ public enum MidpointRounding
+        WinCE = 3,
+        Unix = 4,
+        Xbox = 5,
+        MacOSX = 6
```
Shouldn't this be just be MacOS

stephentoub replied:
This is the existing name.  We can't change it, as part of the goal here is binary compatibility.  In the future we could potentially add another name with the same value.

niemyjski commented on src/System.Runtime.Extensions/ref/System.Runtime.Extensions.cs:
```diff
@@ -534,6 +614,27 @@ public enum MidpointRounding
+    }
+    public enum PlatformID
+    {
+        Win32S = 0,
```
Could we add some comments to these platform id's. I have no clue what Win32S is. Is that Nano?

stephentoub replied:
This is just the enum from desktop:
https://msdn.microsoft.com/en-us/library/3a8hyw88(v=vs.110).aspx

niemyjski commented on src/System.Runtime.Extensions/src/System/Environment.SpecialFolder.cs:
```diff
@@ -0,0 +1,115 @@
+            Resources = SpecialFolderValues.CSIDL_RESOURCES,
+            LocalizedResources = SpecialFolderValues.CSIDL_RESOURCES_LOCALIZED,
+            SystemX86 = SpecialFolderValues.CSIDL_SYSTEMX86,
+            Windows = SpecialFolderValues.CSIDL_WINDOWS,
```
I love that some of this stuff is coming back and it will be easier to port, but I wonder if any of these things make sense... Especially on osx and linux.

stephentoub replied:
Much of it isn't applicable.  But at least for a set of the core libraries there's a goal of enabling binary compatibility, easier porting, etc. with desktop, with Mono, etc., which means values like these come along for the ride.

niemyjski commented:
@stephentoub any ideas when the next release will go out with this?

stephentoub commented:
> any ideas when the next release will go out with this?

I'm not sure.  @richlander or @joshfree may be able to comment.

At the moment it's also in a separate branch, where lots of APIs are being added back, but I expect/hope we'll see that moved over to master soon and then have additional API work happen there.  @danmosemsft may be able to comment on that.

joshfree commented:
> @stephentoub any ideas when the next release will go out with this?

@niemyjski https://github.com/dotnet/corefx/milestones has our most current info
